### ISSUE TITLE
Shadow’s nREPL doesn’t support add-middleware

### ISSUE THREAD

**Issue by tonsky** [thheller/shadow-cljs#1203]
I’m using `add-middleware` in https://github.com/tonsky/Clojure-Sublimed but it seems Shadow version of nREPL doesn’t support it. Can it be added?

I tried specifying `:nrepl {:middleware [nrepl.middleware.dynamic-loader/wrap-dynamic-loader]}` but it doesn’t seem to work:

```
ERROR: Unhandled REPL handler exception processing message {:id 4, :session 7d6799a8-b0d1-437c-aca4-ea3d7ab376fc, :op add-middleware, :middleware [clojure-sublimed.middleware/clone-and-eval clojure-sublimed.middleware/time-eval clojure-sublimed.middleware/wrap-errors clojure-sublimed.middleware/wrap-output], :extra-namespaces [clojure-sublimed.exception clojure-sublimed.middleware]}
clojure.lang.ExceptionInfo: dynamic-loader/*state* is not bond to an atom. This is likely a bug {:state-class nil}
        at nrepl.middleware.dynamic_loader$wrap_dynamic_loader$fn__4613.invoke(dynamic_loader.clj:70)
        at nrepl.middleware$wrap_conj_descriptor$fn__4086.invoke(middleware.clj:16)
        at cider.nrepl$wrap_apropos$fn__34995.invoke(nrepl.clj:187)
        at nrepl.middleware$wrap_conj_descriptor$fn__4086.invoke(middleware.clj:16)
        at nrepl.middleware.lookup$wrap_lookup$fn__5223.invoke(lookup.clj:52)
        at nrepl.middleware$wrap_conj_descriptor$fn__4086.invoke(middleware.clj:16)
```

Looks like this middleware is special and requires special initialization:

https://github.com/nrepl/nrepl/blob/a42eabd309331b8255b06ff562ad3ca9e3034263/src/clojure/nrepl/server.clj#L147-L156

Thank you!

**Comment by thheller** [comment]
Should work as of `2.28.13`. Let me know if not, I didn't really test this in any way. Just upgraded to nrepl 1.3.0 and used the `default-handler` thing.

**Comment by tonsky** [comment]
I think you forget to change arity somewhere

```
[2024-08-26 14:22:54.838 - WARNING] :shadow.cljs.devtools.server.nrepl/middleware-fail - {:sym cider.nrepl/cider-middleware}
ArityException Wrong number of args (2) passed to: shadow.cljs.devtools.server.nrepl/load-middleware
        shadow.cljs.devtools.server.nrepl/load-middleware/fn--31346 (nrepl.clj:85)
        shadow.cljs.devtools.server.nrepl/load-middleware (nrepl.clj:81)
        shadow.cljs.devtools.server.nrepl/load-middleware (nrepl.clj:66)
        shadow.cljs.devtools.server.nrepl/make-middleware-stack (nrepl.clj:105)
        shadow.cljs.devtools.server.nrepl/make-middleware-stack (nrepl.clj:91)
        shadow.cljs.devtools.server.nrepl/start (nrepl.clj:112)
        shadow.cljs.devtools.server.nrepl/start (nrepl.clj:107)
        clojure.lang.Var.invoke (Var.java:386)
        shadow.cljs.devtools.server/start-system/fn--30824 (server.clj:291)
        shadow.cljs.devtools.server/start-system (server.clj:278)
        shadow.cljs.devtools.server/start-system (server.clj:208)
        shadow.cljs.devtools.server/start! (server.clj:537)
```

**Comment by tonsky** [comment]
But `add-middleware` seems to work, thanks!

**Comment by thheller** [comment]
Ah ooops. The `load-middleware` fn is fixed in `2.28.14`.
